### PR TITLE
wg_engine: external webgpu handles (web integration)

### PR DESCRIPTION
**Pull request by SergeyLebedkin**
move instance, adapter and device creation from renderer to application 
its necessary for web integration, because browser have its own mechanics to create hardware handles 
this changes makes webgpu canvas more universal to use in case of system and web applications

See https://github.com/thorvg/thorvg/issues/2410

### CODE REVIEW

File: src/renderer/tvgWgCanvas.cpp [hunk 1]
```diff
@@ -50,11 +50,13 @@ WgCanvas::WgCanvas() : Canvas(nullptr), pImpl(nullptr)
 
 WgCanvas::~WgCanvas()
 {
-    delete pImpl;
+    #ifdef THORVG_WG_RASTER_SUPPORT
+    static_cast<WgRenderer*>(Canvas::pImpl->renderer)->surface = nullptr;
```
Comment (hermet) [inline src/renderer/tvgWgCanvas.cpp:54]: @SergeyLebedkin  Does this really necessary job?  it shouldn't be like this.

Comment (SergeyLebedkin) [reply src/renderer/tvgWgCanvas.cpp:54]: @hermet Yes. We need some how tell to renderer that surface already destroed. If not, it will raise an exception because of Canvas::Impl design: it calls renderer->sync() in disctuctor, but surface is invalid in this stage

```
struct Canvas::Impl
{
    ~Impl()
    {
        //make it sure any deferred jobs
        renderer->sync();

        clearPaints();

        if (renderer->unref() == 0) delete(renderer);
    }

```

File: examples/Example.h [hunk 2]
```diff
@@ -421,6 +423,24 @@ struct WgWindow : Window
...
+
+        // request device
+        const WGPUDeviceDescriptor deviceDesc { .nextInChain = nullptr, .label = "The device", .requiredFeatureCount = featuresCount, .requiredFeatures = featureNames };
+        auto onDeviceRequestEnded = [](WGPURequestDeviceStatus status, WGPUDevice device, char const * message, void * pUserData) { *((WGPUDevice*)pUserData) = device; };
+        wgpuAdapterRequestDevice(adapter, &deviceDesc, onDeviceRequestEnded, &device);
+
+        // device uncaptured error callback
+        auto onDeviceError = [](WGPUErrorType type, char const* message, void* pUserData) { std::cout << message << std::endl; };
+        wgpuDeviceSetUncapturedErrorCallback(device, onDeviceError, nullptr);
+
```
Comment (tinyjin) [inline examples/Example.h:443]: @SergeyLebedkin 
What do you think about we make static helper function to initialize adapter and device?

To remove duplicated code lines and reduce binary size.

I think I'll just copy-paste that codes into WASM binding.

Comment (SergeyLebedkin) [reply examples/Example.h:443]: Yes, we can do this, if it will be hepfull for you

**Review by hermet — CHANGES_REQUESTED** [review]
@SergeyLebedkin user should not destroy a surface, assigned in wg canvas. or they should use target() to unassign before destroying surface.

File: src/renderer/tvgWgCanvas.cpp (hunk 1 above)
Comment (hermet) [reply src/renderer/tvgWgCanvas.cpp:54]: @SergeyLebedkin user should not destroy a surface, assigned in wg canvas. or they should use target() to unassign before destroying surface.

Comment (SergeyLebedkin) [reply src/renderer/tvgWgCanvas.cpp:54]: @hermet I think that we can use terget for un-usigning. Will fix now

File: src/renderer/wg_engine/tvgWgRenderer.cpp [hunk 3]
```diff
@@ -271,14 +272,25 @@ bool WgRenderer::sync()
...
+
+bool WgRenderer::target(WGPUSurface surface, uint32_t w, uint32_t h) {
+        // store target surface properties
+    this->surface = surface;
     mTargetSurface.stride = w;
     mTargetSurface.w = w > 0 ? w : 1;
     mTargetSurface.h = h > 0 ? h : 1;
     
-    mContext.initialize(instance, surface);
+    if (!surface) return false;
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.cpp:293]: If unassignment is also an expected user scenario, it should return true.

File: src/renderer/wg_engine/tvgWgRenderer.cpp [hunk 4]
```diff
@@ -271,14 +272,25 @@ bool WgRenderer::sync()
...
+    initialize();
+    target(surface, w, h);
+    mRenderStoragePool.initialize(mContext, w, h);
+    mStorageRoot.initialize(mContext, w, h);
+    mCompositor.initialize(mContext, w, h);
+    return true;
+}
+
+bool WgRenderer::target(WGPUSurface surface, uint32_t w, uint32_t h) {
+        // store target surface properties
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.cpp:287]: fix the indentation

File: src/renderer/wg_engine/tvgWgRenderer.cpp [hunk 5]
```diff
@@ -245,10 +245,11 @@ bool WgRenderer::clear()
 
 bool WgRenderer::sync()
 {
+    if (!surface) return false;
     disposeObjects();
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.cpp:249]: if the canvas is destroyed in the middle of the usage, no resource leaks?

File: inc/thorvg.h [hunk 6]
```diff
@@ -1754,7 +1754,7 @@ class TVG_API WgCanvas final : public Canvas
      * @see Canvas::viewport()
      * @see Canvas::sync()
      */
-    Result target(void* instance, void* surface, uint32_t w, uint32_t h) noexcept;
+    Result target(void* instance, void* adapter, void* device, void* surface, uint32_t w, uint32_t h) noexcept;
```
Comment (hermet) [inline inc/thorvg.h:1757]: please update the api doc.

File: examples/Example.h (hunk 2 above)
Comment (hermet) [reply examples/Example.h:427]: @SergeyLebedkin Please merge `createHandles()`; into `WgWindow()`.

**Review by hermet — CHANGES_REQUESTED** [review]

File: src/renderer/wg_engine/tvgWgRenderer.cpp (hunk 5 above)
Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgRenderer.cpp:248]: Objects will be destoyed when renderer will be released.
But we can remove them in any case

**Comment by SergeyLebedkin** [comment]
@hermet 
> Please merge createHandles(); into WgWindow().
Its already in WgWindow class, what do you mean "merge"

**Comment by hermet** [comment]
> @hermet
> 
> > Please merge createHandles(); into WgWindow().
> > Its already in WgWindow class, what do you mean "merge"

@SergeyLebedkin Merge the code from createHandle() into WgWindow(). Rather than adding function call.

File: src/renderer/tvgWgCanvas.cpp [hunk 7]
```diff
@@ -50,11 +50,14 @@ WgCanvas::WgCanvas() : Canvas(nullptr), pImpl(nullptr)
 
 WgCanvas::~WgCanvas()
 {
-    delete pImpl;
+    #ifdef THORVG_WG_RASTER_SUPPORT
```
Comment (hermet) [inline src/renderer/tvgWgCanvas.cpp:53]: please remove the indentation for `#ifdef ~ #endif `

File: examples/Example.h [hunk 8]
```diff
@@ -435,20 +455,23 @@ struct WgWindow : Window
...
         //Set the canvas target and draw on it.
-        verify(canvas->target(instance, surface, width, height));
+        verify(canvas->target(instance, adapter, device, surface, width, height));
     }
 
     void refresh() override 
     {
         wgpuSurfacePresent(surface);
     }
+
```
Comment (hermet) [inline examples/Example.h:474]: -- unnecessary change

File: src/renderer/wg_engine/tvgWgRenderer.h [hunk 9]
```diff
@@ -55,13 +56,15 @@ class WgRenderer : public RenderMethod
...
+    WGPUSurface surface{}; // external handle
+
 private:
     WgRenderer();
     ~WgRenderer();
     void initialize();
     void release();
     void disposeObjects();
-
+
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.h:67]: -- white spaces

File: inc/thorvg.h [hunk 10]
```diff
@@ -1754,7 +1756,7 @@ class TVG_API WgCanvas final : public Canvas
...
```
Comment (hermet) [inline inc/thorvg.h:1759]: Please double-check the `adapter` parameter here. I don't see any necessity, otherwise please remove it.
```
Result target(void* instance, void* surface, uint32_t w, uint32_t h, void* device = nullptr)
```

File: inc/thorvg.h [hunk 11]
```diff
@@ -1742,6 +1742,8 @@ class TVG_API WgCanvas final : public Canvas
      * @brief Sets the drawing target for the rasterization.
      *
      * @param[in] instance WGPUInstance, context for all other wgpu objects.
+     * @param[in] adapter WGPUAdapter, handle for webgpu adapter
+     * @param[in] device WGPUDevice, handle for webgpu device
```
Comment (hermet) [inline inc/thorvg.h:1746]: ```@param[in] device WGPUDevice: desired handle for the WebGPU device. If it is nullptr, ThorVG will assign an appropriate device internally.```

**Review by hermet — CHANGES_REQUESTED** [review]
@SergeyLebedkin Let's roll back this part. I think we can handle it in the WG renderer for user convenience, inheriting only when the device value is passed.
```
        // request adapter
        const WGPURequestAdapterOptions requestAdapterOptions { .nextInChain = nullptr, .compatibleSurface = surface, .powerPreference = WGPUPowerPreference_HighPerformance, .forceFallbackAdapter = false };
        auto onAdapterRequestEnded = [](WGPURequestAdapterStatus status, WGPUAdapter adapter, char const * message, void * pUserData) { *((WGPUAdapter*)pUserData) = adapter; };
        wgpuInstanceRequestAdapter(instance, &requestAdapterOptions, onAdapterRequestEnded, &adapter);

        // get adapter and surface properties
        WGPUFeatureName featureNames[32]{};
        size_t featuresCount = wgpuAdapterEnumerateFeatures(adapter, featureNames);

        // request device
        const WGPUDeviceDescriptor deviceDesc { .nextInChain = nullptr, .label = "The device", .requiredFeatureCount = featuresCount, .requiredFeatures = featureNames };
        auto onDeviceRequestEnded = [](WGPURequestDeviceStatus status, WGPUDevice device, char const * message, void * pUserData) { *((WGPUDevice*)pUserData) = device; };
        wgpuAdapterRequestDevice(adapter, &deviceDesc, onDeviceRequestEnded, &device);

        // device uncaptured error callback
        auto onDeviceError = [](WGPUErrorType type, char const* message, void* pUserData) { std::cout << message << std::endl; };
        wgpuDeviceSetUncapturedErrorCallback(device, onDeviceError, nullptr);
```

File: src/renderer/wg_engine/tvgWgRenderer.h [hunk 12]
```diff
@@ -74,7 +77,7 @@ class WgRenderer : public RenderMethod
     WgContext mContext;
     WgPipelines mPipelines;
     WgCompositor mCompositor;
-
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.h:77]: please remove white spaces.

File: src/renderer/wg_engine/tvgWgRenderer.cpp [hunk 13]
```diff
@@ -271,14 +272,43 @@ bool WgRenderer::sync()
 
 
 // target for native window handle
-bool WgRenderer::target(WGPUInstance instance, WGPUSurface surface, uint32_t w, uint32_t h)
+bool WgRenderer::target(WGPUInstance instance, WGPUSurface surface, uint32_t w, uint32_t h, WGPUDevice device)
 {
+    WGPUDevice dev = device;
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.cpp:277]: please make it code simple. i think you can directly use `device`.

File: src/renderer/tvgWgCanvas.cpp [hunk 14]
```diff
@@ -50,11 +50,14 @@ WgCanvas::WgCanvas() : Canvas(nullptr), pImpl(nullptr)
 
 WgCanvas::~WgCanvas()
 {
-    delete pImpl;
+#ifdef THORVG_WG_RASTER_SUPPORT
+    auto renderer = static_cast<WgRenderer*>(Canvas::pImpl->renderer);
+    renderer->target(nullptr, 1, 1);
```
Comment (hermet) [inline src/renderer/tvgWgCanvas.cpp:55]: please pass 0, 0

File: src/renderer/wg_engine/tvgWgRenderer.cpp [hunk 15]
```diff
@@ -271,14 +272,43 @@ bool WgRenderer::sync()
...
+    mStorageRoot.initialize(mContext, w, h);
+    mCompositor.initialize(mContext, w, h);
+    return true;
+}
+
+bool WgRenderer::target(WGPUSurface surface, uint32_t w, uint32_t h) {
     // store target surface properties
+    this->surface = surface;
     mTargetSurface.stride = w;
     mTargetSurface.w = w > 0 ? w : 1;
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.cpp:308]: I think you should not change w/h value here.

File: examples/Example.h [hunk 16]
```diff
@@ -421,6 +423,24 @@ struct WgWindow : Window
         surfaceDesc.label = "The surface";
         surface = wgpuInstanceCreateSurface(instance, &surfaceDesc);
 
+        // request adapter
```
Comment (hermet) [inline examples/Example.h:426]: no need to change Example code anymore.

File: examples/Example.h (hunk 2 above)
Comment (hermet) [reply examples/Example.h:443]: @SergeyLebedkin please revert the example code all here and just pass null for device.

**Review by hermet — CHANGES_REQUESTED** [review]

**Review by hermet — APPROVED** [review]

**Review by tinyjin — APPROVED** [review]
Confirmed that it works in the browser
